File: rs/minsk/src/analysis/syntax/asserting_enumerator.rs
use super::kind::SyntaxKind;
use super::node::SyntaxNode;

pub(crate) struct AssertingEnumerator<'t> {
    enumerator: Enumerator<'t>,
}

impl<'t> AssertingEnumerator<'t> {
    pub(crate) fn new(node: &'t dyn SyntaxNode) -> Self {
        Self {
            enumerator: Enumerator::new(node),
        }
    }

    pub(crate) fn assert_node(&mut self, kind: SyntaxKind) {
        let current = self.enumerator.next().unwrap();
        assert_eq!(kind, current.kind());
        assert!(!current.is_token());
    }

    pub(crate) fn assert_token(&mut self, kind: SyntaxKind, text: &str) {
        let current = self.enumerator.next().unwrap();
        assert_eq!(kind, current.kind());
        assert!(current.is_token());
        assert_eq!(text, current.token_text());
    }

    pub(crate) fn assert_at_end(&mut self) {
        assert!(self.enumerator.next().is_none());
    }
}

struct Enumerator<'t> {
    stack: Vec<&'t dyn SyntaxNode>,
}

impl<'t> Enumerator<'t> {
    fn new(node: &'t dyn SyntaxNode) -> Self {
        Self { stack: vec![node] }
    }
}

impl<'t> Iterator for Enumerator<'t> {
    type Item = &'t dyn SyntaxNode;

    fn next(&mut self) -> Option<Self::Item> {
        let n = match self.stack.pop() {
            Some(n) => n,
            None => return None,
        };
        for child in n.children().into_iter().rev() {
            self.stack.push(child);
        }
        Some(n)
    }
}

File: nim/tests/assertingEnumerator.nim
import std/algorithm
import std/options
import std/strutils

import minskpkg/codeAnalysis/syntax/[
  syntaxKind,
  syntaxNode,
  syntaxToken,
]

type
  Enumerator = object
    nodes: seq[SyntaxNode]
    i: int

  AssertingEnumerator* = object
    enumerator: Enumerator
    node: SyntaxNode
    failed: bool

proc newEnumerator(nodes: sink seq[SyntaxNode]): Enumerator =
  result.nodes = nodes
  result.i = 0

proc next(e: var Enumerator): Option[SyntaxNode] =
  for n in e.nodes:
    echo n.kind
  echo ""
  if e.i < e.nodes.len:
    result = some(e.nodes[e.i])
    e.i += 1

proc flatten(node: SyntaxNode): seq[SyntaxNode] =
  var stack: seq[SyntaxNode] = @[]
  stack.add(node)
  while stack.len > 0:
    let current = stack.pop()
    result.add(current)
    var children = current.children
    children.reverse()
    for child in children:
      stack.add(child)

proc newAssertingEnumerator*(node: SyntaxNode): AssertingEnumerator =
  result.enumerator = newEnumerator(flatten(node))
  result.node = node
  result.failed = false

proc assertToken*(e: var AssertingEnumerator, kind: SyntaxKind, text: string) =
  try:
    let n = e.enumerator.next().get
    assert ($n.kind).endsWith("Token")
    let token = n.SyntaxToken
    assert token.kind == kind
    assert token.text == text
  except AssertionDefect:
    e.failed = true
    raise

proc assertNode*(e: var AssertingEnumerator, kind: SyntaxKind) =
  try:
    let n = e.enumerator.next().get
    assert n isnot SyntaxToken
    assert n.kind == kind
  except AssertionDefect:
    e.failed = true
    raise

proc `=destroy`*(e: var AssertingEnumerator) =
  if not e.failed:
    assert e.enumerator.next().isNone
